designs/2024-deprecated-rule-metadata/README.md

```diff
@@ -0,0 +1,281 @@
+- Some rules provide the replacement rule name with the plugin prefix as in `prefix/rule-name` while others just provide it as `rule-name`, which can result in ambiguity about whether the replacement rule is in the same plugin, a different third-party plugin, or ESLint core. And for third-party plugins, there's no easy way to automatically determine where their documentation is located or how to link to them.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+We propose to extend `meta.deprecated` rule property schemas to reduce ambiguity and allow additional key details to be represented, described below using TypeScript types for clarity:
+
+```ts
+type RuleMeta = {
+  deprecated?:
+    | boolean // Existing boolean option, backwards compatible.
+    | string // Shorthand property for general deprecation message, such as why the deprecation occurred. Empty strings are forbidden as they are falsy.
+    | DeprecatedInfo // Proposed extension
+
+  /** @deprecated */
+  replacedBy?: string[] // Deprecate the top-level property and "move" into the "deprecated" object
+};
+
+/* At least one property is required */
+type DeprecateInfo = {
+  info?: Message
+  replacedBy?: (string|ReplacedByInfo)[] // An empty array explicitly states that there is no replacement
+}
+
+/* At least one property is required */
+type ReplacedByInfo = {
+  plugin?: string | Specifier // name should be "eslint" if the replacemenet is an ESLint core rule. Omit the property if the replacement is in the same plugin
+  rule?: string | Specifier
+  info?: string | Message
+  kind?: ReplacementKind // Defaults to "moved" if missing
+  deprecatedSince?: Version // Helps users gauge when to migrate and useful for documentation
+  availableUntil?: Version | null // The estimated version when the rule is removed (probably the next major version). null means the rule is "frozen" (will be available but will not be changed)
+}
+
+type Message = {
+  message: string // General message presented to the user. Content depends on the property (e.g. for the key rule why the rule is deprecated or for info how to replace the rule)
+  url:            // URL to more information about this deprecation in general.
+}
+
+type Specifier = {
+  name: string // Name of the rule / configuration / ...
+  url: string // URL to more information about this deprecation in general.
+}
+
+type ReplacementKind =
+  'moved' |  // The rule has moved to another plugin if plugin is set, otherwise the rule is renamed in the same plugin
+  'merged' | // The rule merged with another rule
+  'option'   // The current rule behavior is available as an option in the replacement rule
```

Comment:
I think `ReplacedByInfo.kind` isn't necessary as this info and more context can be provided as part of the `ReplacedByInfo.info` (or ReplacedByInfo.info.message) text.

Reply:
The idea behind this property was:
- for documentation generators: able to provide a default "message" without having to specify the same message for multiple rules (e.g. the rule was renamed from "foo" to "bar")
- a possible config migration tool: suggest changes without relying on a specific format in info.message (e.g. rename the rule in the config)

Reply:
I'm also not sure of the value here. A single word just doesn't given enough context. I also fear this would become a maintenance issue as we'll end finding other reasons rules were deprecated and need to update this list with more unclear terms.

Reply:
Looks like we're in favor of removing this property. @DMartens can you update the RFC?